src/Cardano/Wallet/Network.hs

```diff
@@ -73,27 +73,49 @@ data TickResult a
 -- | Retrieve blocks from a chain producer and execute some given action for
 -- each block.
 listen
-    :: forall e0 e1. (Exception e0)
+    :: forall a e0 e1. (Exception e0)
     => NetworkLayer IO e0 e1
     -> ([Block] -> IO ())
+    -> IO a
+listen network action =
+    tick (getNextBlocks network) action 5000 (SlotId 0 0)
+
+getNextBlocks
+    :: forall e0 e1. (Exception e0)
+    => NetworkLayer IO e0 e1
+    -> SlotId
+    -> IO (TickResult [Block], SlotId)
+getNextBlocks network current = do
+    res <- runExceptT $ nextBlocks network current
+    -- NOTE
+    -- In order to avoid having to perform some slotting arithmetic, we only
+    -- process blocks if we receive more than one, such that we can use the
+    -- last block as the starting point for the next batch.
+    -- The trade-off is is that we'll be fetching this last block twice,
+    -- which is fair price to pay in order NOT to have to do any slotting
+    -- arithmetic nor track how many blocks are present per epochs.
+    case res of
+        Left err -> throwIO err
+        Right bs | length bs < 2 ->
+            pure (Sleep, current)
+        Right blocks ->
+            let next = slotId . header . last $ blocks
+            in pure (GotChunk (init blocks), next)
+
+-- | Process all available blocks from the network layer with the given action.
+-- When the network layer reports no more blocks, stop.
+drain
+    :: forall e0 e1. (Exception e0)
+    => NetworkLayer IO e0 e1
+    -> (SlotId -> [Block] -> IO ())
     -> IO ()
-listen network action = do
-    tick getNextBlocks action 5000 (SlotId 0 0)
+drain network action = go (SlotId 0 0)
   where
-    getNextBlocks :: SlotId -> IO (TickResult [Block], SlotId)
-    getNextBlocks current = do
-        res <- runExceptT $ nextBlocks network current
-        -- NOTE
-        -- In order to avoid having to perform some slotting arithmetic, we only
-        -- process blocks if we receive more than one, such that we can use the
-        -- last block as the starting point for the next batch.
-        -- The trade-off is is that we'll be fetching this last block twice,
-        -- which is fair price to pay in order NOT to have to do any slotting
-        -- arithmetic nor track how many blocks are present per epochs.
+    go :: SlotId -> IO ()
+    go start = do
+        (res, start') <- getNextBlocks network start
         case res of
-            Left err -> throwIO err
-            Right bs | length bs < 2 ->
-                pure (Sleep, current)
-            Right blocks ->
-                let next = slotId . header . last $ blocks
-                in pure (GotChunk (init blocks), next)
+            GotChunk blocks -> do
+                action start' blocks
+                go start'
+            Sleep -> pure ()
```

Comment:
The more we make changes to these interfaces, the more it makes me believe that we are doing something wrong and we should be defining functions in the network layer interface instead. I believe, everything we need could be achieved with a simple change in the network interface:

```hs
nextBlocks :: SlotId -> ExceptT e0 m Result

data Result
    = NextBlocks !([Blocks], SlotId) -- ^ Have a result, a batch of next blocks and, the next Id to start from
    | NoBlocks -- ^ There is no blocks available now, so wait.
```

This way, the chain producer implementation is charge of letting others know about what is the next slot id and, we can get rid of `listen`, `getNextBlocks` and `drain` in favor of an approach that is also probably more resilient to the upcoming changes in the network protocol with Shelley.

Reply:
Sure - where do you envision the "listen" process sitting in relation to wallet layer and network layer?

Reply:
I'll make a proposal in another PR, it will be simpler to discuss around that :+1: